### DISCUSSION TITLE
Simple drawing on SVG

### DISCUSSION THREAD

**Discussion by duboism** [ImageMagick/ImageMagick#7745] [Help]
### ImageMagick version

7.1.1-39

### Operating system, version and so on

Fedora 40

### Description

Hello,

I'm working on a PHP application where we use ImageMagick to draw some image either in PNG or in SVG. The image is not hard: a few `polygon`s and `annotation`s. Everything works with PNG but SVG export is useless. If I remove the `annotation`s, it's a bit better but not great (in particular, the `polygon`s are not filled correctly). 

In order to clarify, here are some examples on the command-line:
  - `magick -size 100x60 xc:transparent -fill white -stroke black -draw "rectangle 20,10 80,50" draw_rect.svg`: incorrect filling
  - `magick -size 100x60 xc:transparent -fill white -stroke black -draw "rectangle 20,10 80,50" -draw 'text 10,10 "Hello"' draw_rect_text.svg`: output just a path looking like an "o" (no rectangle)
Using PNG as output works as expected.

Am I doing something wrong ?

Thanks in advance

**Comment by snibgo** [comment]
ImageMagick is a raster image processor. It also contains some vector operations, but it isn't a vector image processor, and it isn't an SVG editor.

Operations like `-draw "rectangle 20,10 80,50"` will draw a rectangle on a raster image.

**Reply by duboism** [reply]
I agree that IM is more raster oriented but the commands more or less work and there is a whole section about drawing in the manual that talks about SVG ([this section](https://usage.imagemagick.org/draw/)). Finally, despite my limited knowledge of MVG, I was able to create an MVG file more or less corresponding to this test and convert it to a correct SVG.

So I think there is a way.

**Reply by gammalogic** [reply]
> ImageMagick is a raster image processor. It also contains some vector operations, but it isn't a vector image processor, and it isn't an SVG editor.

If ImageMagick internally saves the state of vector drawing operations it would be good to have some way to get the data back out again as an SVG. I'm working on a drawing application at the moment using the Imagick extension in PHP and want to give users the option of PNG and/or SVG export, but I ended up having to create new routines that duplicate the ImageMagick routines so I can then build the SVG file - essentially, every time I tell ImageMagick to draw something I have to call my own routine at the same time to save the curve/polygon data and then generate the correct tags at the end. A script to convert the data generated by the `getVectorGraphics` function (using the Imagick extension) would be another way to do this, but I went for what seemed like the easier option as I wasn't familiar with the format generated by ImageMagick.

**Reply by dlemstra** [reply]
>ImageMagick is a raster image processor.

This means that we store the state as a raster image and not as a vector image.

**Reply by gammalogic** [reply]
Where does the data generated by the `getVectorGraphics` function come from? When ImageMagick draws a curve, it must still save the drawing operations somewhere before rasterizing the image? Or is this only a function of the Imagick extension?

**Reply by dlemstra** [reply]
We don't maintain IMagick so we cannot give you an answer to that question.

**Reply by duboism** [reply]
> This means that we store the state as a raster image and not as a vector image.

I'm not sure I understand but isn't the fact that you can generate an SVG from an MVG contradicts this ?

**Reply by dlemstra** [reply]
> I'm not sure I understand but isn't the fact that you can generate an SVG from an MVG contradicts this ?

The fact that we store an image as a raster image is not up for debate 😁Our internal image format is a raster image.

**Reply by duboism** [reply]
OK, sorry, I didn't mean to doubt what you were saying but to understand. By the way I took some time to look at the code of IMagick's `getVectorData` and I think that it uses the C function `DrawGetVectorGraphics` which is related to the [MagickWand API](https://imagemagick.org/api/drawing-wand.php). Not sure it helps.

**Reply by dlemstra** [reply]
We have drawing operations and in theory you might be able to change the output of `DrawGetVectorGraphics` into a vector image but that all depends on what you are doing.

**Reply by duboism** [reply]
Thanks. As I said, I just need some polygons and a bit of text. As the CLI examples show, rectangles seems more or less OK but text is useless. Can you provide a working CLI example? From there, I might be able to translate it to PHP.

**Reply by dlemstra** [reply]
A working CLI example of what?

**Reply by duboism** [reply]
I'm looking for a working example of a SVG file with a filled rectangle and some text (similar to those in the very first message of this discussion). If you have other enlightening examples don't hesitate :).

**Reply by gammalogic** [reply]
If I take the example on [this page](https://usage.imagemagick.org/draw/#settings) and change the output to SVG

`
magick -size 100x60 xc:skyblue -draw "fill green circle 41,39 44,57 fill blue circle 59,39 56,57 fill red circle 50,21 50,3" draw_circle_multi.svg
` 

it is obvious that the SVG file is not well formed

```xml
<?xml version="1.0" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 20010904//EN"
  "http://www.w3.org/TR/2001/REC-SVG-20010904/DTD/svg10.dtd">
<svg width="100" height="60">
fill:green;  <circle cx="41" cy="39" r="18.2483"/>
fill:blue;  <circle cx="59" cy="39" r="18.2483"/>
fill:red;  <circle cx="50" cy="21" r="18"/>
</svg>
```

So, it appears that ImageMagick allows SVG graphics to be generated but the data is wrong inside of it. If ImageMagick does not support SVG, why does it try to export to it?

**Reply by duboism** [reply]
Maybe I was not clear but this is the problem I tried to describe (it's even worse if you use `annotate`). Thanks for the detailed example, I should have been more clear in my explanation (I was not sure if the problem was specific to PHP or not).

**Reply by duboism** [reply]
Sorry to resuscitate the thread but should we consider the (great) example given by @gammalogic as a bug ?

**Reply by dlemstra** [reply]
The svg coder appears to do something "smart" by looking at the last draw operation and then use that information to create an svg image. But looking at the implementation and the history of the file it looks like this never worked correctly. 

Fixing this would require a complete rewrite of that code and I don't think we will do that anytime soon because it has been broken for over 15 years and you seem to be the first one noticing this. I wasn't even aware we did this and saw this part of our code for the first time today. We of course welcome contributions from other people so I will mark this discussion as up for grabs but I would advise you to find a different solution for this now because I don't think this will get fixed anytime soon.

**Reply by duboism** [reply]
Thanks for you answer. Could you point to the issue in question?

**Reply by dlemstra** [reply]
We will not be creating an issue. I have labeled this as up for grabs and as an enhancement so people know they can work on it.

**Reply by duboism** [reply]
Ah OK, I didn't understand what you meant. It's probably too hard for me but is there some doc I  (or another person in the thread) could read to try to improve the situation?

**Reply by dlemstra** [reply]
Someone would need to rewrite the code in the svg writer there is not really a document on what they should do. They would need to read the svg spec and figure out how to write everything correctly.

**Reply by duboism** [reply]
Sure. But I was also thinking about learning the internal of ImageMagick's development. I took a look at `coders/svg.{h, c}` and they are full of ImageMagick dialect (thinks like `AcquireUniqueFileResource`,  etc). Or for instance where to start, what patterns are used, etc.

**Reply by dlemstra** [reply]
There is no documentation for developers. This will be going to be a lot of self study by looking at the existing code.

**Comment by snibgo** [comment]
I have looked at this briefly. The problem where...
```
fill:green;  <circle cx="41" cy="39" r="18.2483"/>
```
... should be ...
```
<circle fill="green" cx="41" cy="39" r="18.2483"/>
```
... is probably fairly easy to fix. Build a stack of object attributes (like "fill") as we come across them. When we know the object (eg "circle"), pop the attributes (maybe FIFO order?) and emit them into the SVG.

However, ensuring this fix doesn't cause other problems would not be simple. And there are probably many other instances where IM generates bad SVG, or at least SVG that doesn't accurately do what is needed.

I agree with @dlemstra that this needs "a complete rewrite of that code". I suggest that this could start by collecting examples of `-draw` operations, and the SVG that IM currently generates, and the SVG that IM should generate. That would form a test suite. Once that is done, code can be designed and written.

This would not be trivial. I would guess it would take a month or three. I am not volunteering.

**Reply by gammalogic** [reply]
> I agree with @dlemstra that this needs "a complete rewrite of that code". I suggest that this could start by collecting examples of `-draw` operations, and the SVG that IM currently generates, and the SVG that IM should generate. That would form a test suite. Once that is done, code can be designed and written.

I noticed also that some operations were ignored. In my command line example, the xc:skyblue attribute was ignored and the background was not filled.

**Reply by duboism** [reply]
Thanks for your thoughts on that. Given @dlemstra comment, I guess this would take me considerable time to get this. Could it be part of a Google Summer of Code project or other mentoring program?

**Reply by dlemstra** [reply]
That would also require a big investment from our side and I don't think that will happen anytime soon.

**Comment by snibgo** [comment]
>  ... the xc:skyblue attribute was ignored and the background was not filled.

Yes. And that raises the question: should an entire IM command be converted to SVG? If so, then the problem expands from that of a single `-draw` operation.

But this also raises the question: Why is someone writing an IM command that is then to be translated to SVG? Why not instead directly write SVG?

**Reply by dlemstra** [reply]
I even wonder if we should just remove this broken feature that hasn't be working properly for ages.

**Reply by duboism** [reply]
> But this also raises the question: Why is someone writing an IM command that is then to be translated to SVG? Why not instead directly write SVG?

I'm not sure that I understand what you mean but in my case, we want to draw some images in either PNG or SVG formats (remember we use `IMagick` PHP extension). In fact, I wanted to use `IMagick` because the library we currently use ([Image_Canvas](https://github.com/pear/Image_Canvas)) doesn't work on PHP8 (despite my efforts).

**Reply by duboism** [reply]
> I even wonder if we should just remove this broken feature that hasn't be working properly for ages.

I feared we reach this conclusion. Do you talk about all SVG export ?

**Comment by snibgo** [comment]
> ...  we want to draw some images in either PNG or SVG formats

You could generate SVG, then rasterize that to PNG.

**Reply by duboism** [reply]
Your suggestion is interesting and I'm sure we will find a solution for our case. It's just that I expected the same code based on `IMagick` to generate both formats, just changing the output option as was done with [Image_Canvas](https://github.com/pear/Image_Canvas/) (also conversion from Image_Canvas was pretty easy).

**Comment by snibgo** [comment]
I wrote:
> And that raises the question: should an entire IM command be converted to SVG?

Entire IM commands could be converted to SVG, but that task is very difficult. For example: an SVG can contain embedded raster images. Objects (such as raster images, or vector objects such as lines and areas) can have "filter" operations, such as blurs. Hence many IM operations such as `-blur` have an analogue in SVG, so a new translator for IM commands would need to be written. This would be an architectural problem, as IM works by doing an operation and forgetting about it, then doing the next operation, etc. Only when an image is written does IM know that it needs to translate all the forgotten operations to SVG. And the translation in ambiguous: if an IM command draws a line and then blurs it, what gets written to SVG, either a vector line with a blur filter, or a raster image of the blurred line?

No, I conclude it is not feasible to convert entire IM commands to SVG.

The lesser goal, of correctly translating single `-draw` operations to SVG is feasible, I think. But not trivial.

**Comment by fmw42** [comment]
Perhaps see https://usage.imagemagick.org/draw/#paths and https://usage.imagemagick.org/draw/#svg_output

**Reply by duboism** [reply]
Thanks for your answer but I don't understand what you mean. It's quite clear from the discussion that SVG output (at least fro draw commands) is buggy.  I don't see how the links you sent (which I read before posting) would help. Can you elaborate ?

**Reply by fmw42** [reply]
Just suggesting that you might just create your own commands in SVG in Imagemagick to do your drawing.  Might not be relevant.

**Reply by duboism** [reply]
I know what I want to draw and it works on PNG (by the way, the PNG file is much better than with GD). The problem is that the same commands on a SVG file leads to an unusable file. So I guess your suggestion is not relevant for my problem.
